### PR TITLE
CI: build build-tools image

### PR DESCRIPTION
**Pull request by abhigets**
## Summary of changes
We had the docker file that we use to build our binary and other tools resided in the build repo
It made sense to bring the docker file to its repo where it has been used
So that the contributors can also view it and amend if required
It will reduce the maintenance. Docker file changes and code changes can be accommodated in same PR
Also, building the image and pushing it to ECR is been abstracted in a reusable workflow. Ideal is to use that for any other jobs too

## Checklist before requesting a review

- [x] Moved the docker file used to build the binary from the build repo to the neon repo
- [x] adding gh workflow to build and push the image
- [x] adding gh workflow to tag the pushed image
- [x] update readMe file

### CODE REVIEW

**Review by bayandin — COMMENTED** [review]
Great start!
Added a bunch of comments, but most of them related to codestyle/naming

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -63,22 +63,56 @@ jobs:
...
+      - name: Checkout Repository
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Check if Dockerfile.buildtools has changed
+        id: dockerfile
+        run: |
+          updated_files=$(git diff --name-only ${{ github.event.before }} ${{ github.event.pull_request.base.sha }})
+          echo "$updated_files"
```
[the remark this answers is not in the document]
Comment (abhigets) [reply .github/workflows/build_and_test.yml]: done removed

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -63,22 +63,56 @@ jobs:
           echo run:$GITHUB_RUN_ID
           echo ref:$GITHUB_REF_NAME
           echo rev:$(git rev-list --count HEAD)
+          echo "image_tag=pinned" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (abhigets) [reply .github/workflows/build_and_test.yml:66]: Done .. Pushed the changes

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -63,22 +63,56 @@ jobs:
...
+
+      - name: Checkout Repository
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Check if Dockerfile.buildtools has changed
+        id: dockerfile
+        run: |
+          updated_files=$(git diff --name-only ${{ github.event.before }} ${{ github.event.pull_request.base.sha }})
```
[the remark this answers is not in the document]
Comment (abhigets) [reply .github/workflows/build_and_test.yml]: failed to run git: fatal: not a git repository (or any of the parent directories): .git
it seems for this to work we need to clone the repo

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -107,11 +141,11 @@ jobs:
       - name: Run mypy to check types
         run: poetry run mypy .
 
-  check-codestyle-rust:
-    needs: [ check-permissions ]
+  check-codestyle-build-tools:
```
[the remark this answers is not in the document]
Comment (abhigets) [reply .github/workflows/build_and_test.yml:151]: yep good catch

File: .github/workflows/build_and_test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bayandin) [reply .github/workflows/build_and_test.yml:66]: We don't need these changes, we already have in the next step

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -63,22 +63,59 @@ jobs:
...
         shell: bash
         id: build-tag
 
-  check-codestyle-python:
+      - name: Get buildtools tag
+        run: |
+          echo "image_tag=pinned" >> $GITHUB_OUTPUT
+          if [[ "$GITHUB_REF_NAME" != "main" && "$GITHUB_REF_NAME" != "release" ]]; then
+            echo "image_tag=$GITHUB_RUN_ID" >> $GITHUB_OUTPUT
+          fi
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:84]:
```suggestion
          if [[ "$GITHUB_EVENT_NAME" == "pull_request" ]]; then
            IMAGE_TAG=$GITHUB_RUN_ID
          else
            IMAGE_TAG=pinned
          fi
          
          echo "image_tag=${IMAGE_TAG}" >> $GITHUB_OUTPUT
```

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -63,22 +63,59 @@ jobs:
...
+    steps:
+      - name: Check if Dockerfile.buildtools has changed
+        id: dockerfile
+        run: |
+          updated_files=$(gh pr --repo neondatabase/neon diff ${{ github.event.pull_request.number }} --name-only)
+          if [[ $updated_files == *"Dockerfile.buildtools"* ]]; then
+            echo "docker_file_changed=true" >> $GITHUB_OUTPUT
+          fi
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:102]:
```suggestion
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

File: .github/workflows/build_and_test.yml [hunk 7]
```diff
@@ -63,22 +63,56 @@ jobs:
...
+      docker-file-path: Dockerfile.buildtools
+      image-name: build-tools
+    secrets: inherit
+
+  check-codestyle-python:
+    needs: [ check-permissions, push-buildtools-docker-file , tag ]
+    runs-on: [ self-hosted, gen3, small ]
     container:
-      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/rust:pinned
+      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{  needs.tag.outputs.build-tools-tag }}
```
[the remark this answers is not in the document]
Comment (bayandin) [reply .github/workflows/build_and_test.yml:118]: And these changes got lost 😄

File: .github/workflows/build_and_test.yml [hunk 8]
```diff
@@ -479,12 +516,12 @@ jobs:
       # while coverage is currently collected for the debug ones
 
   create-test-report:
-    needs: [ check-permissions, regress-tests, coverage-report, benchmarks ]
+    needs: [ check-permissions, regress-tests, coverage-report, benchmarks, build-buildtools-image , tag]
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:519]:
```suggestion
    needs: [ check-permissions, regress-tests, coverage-report, benchmarks, build-buildtools-image, tag ]
```

File: .github/workflows/update_build-tools-image.yml [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+            echo "prev-digest-buildtools=$PREV_DIGEST" >> $GITHUB_OUTPUT
+          fi
+
+      - name: Tag image
+        run: |
+          crane tag "${IMAGE}:${FROM_TAG}" "${TO_TAG}"
+
+  rollback-tag-image:
+    needs:  tag-image
```
Comment (bayandin) [inline .github/workflows/update_build-tools-image.yml:83]:
```suggestion
    needs: tag-image
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -70,3 +70,17 @@ We're using the following approach to make it work:
...
+## How do I add the "pinned" tag to an buildtools image?
+We use the `pinned` tag for `Dockerfile.buildtools` build images in our CI/CD setup, currently adding the `pinned` tag is a manual operation.
+
+You can call it from GitHub UI: https://github.com/neondatabase/neon/actions/workflows/update_build-tools-image.yml,
+or using GitHub CLI:
+
+```bash
+gh workflow -R neondatabase/neon run update_build-tools-image.yml \
+            -f from-tag=6254913013 \
+            -f to-tag=mynewtag \
```
Comment (bayandin) [inline CONTRIBUTING.md:83]:
```suggestion
            -f to-tag=pinned \
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -70,3 +70,17 @@ We're using the following approach to make it work:
...
 For details see [`approved-for-ci-run.yml`](.github/workflows/approved-for-ci-run.yml)
+
+## How do I add the "pinned" tag to an buildtools image?
+We use the `pinned` tag for `Dockerfile.buildtools` build images in our CI/CD setup, currently adding the `pinned` tag is a manual operation.
+
+You can call it from GitHub UI: https://github.com/neondatabase/neon/actions/workflows/update_build-tools-image.yml,
+or using GitHub CLI:
+
+```bash
+gh workflow -R neondatabase/neon run update_build-tools-image.yml \
```
Comment (bayandin) [inline CONTRIBUTING.md:81]:
```suggestion
gh workflow -R neondatabase/neon run update_build_tools_image.yml \
```

File: .github/workflows/update_build-tools-image.yml
Comment (bayandin) [inline .github/workflows/update_build-tools-image.yml:1]: Let's rename the file to `.github/workflows/update_build_tools_image.yml` to have consistent naming with other workflows

File: .github/workflows/update_build-tools-image.yml [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+      from-tag:
+        description: 'Source tag'
+        required: true
+        type: string
+      to-tag:
+        description: 'Destination tag'
+        required: true
+        type: string
+        default: 'pinned'
+  pull_request:
```
Comment (bayandin) [inline .github/workflows/update_build-tools-image.yml:18]: We don't want to run this on pull requests automatically
```suggestion
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -70,3 +70,17 @@ We're using the following approach to make it work:
 - The label gets removed automatically, so to run CI again with new changes, the label should be added again (after the review)
 
 For details see [`approved-for-ci-run.yml`](.github/workflows/approved-for-ci-run.yml)
+
+## How do I add the "pinned" tag to an buildtools image?
+We use the `pinned` tag for `Dockerfile.buildtools` build images in our CI/CD setup, currently adding the `pinned` tag is a manual operation.
+
+You can call it from GitHub UI: https://github.com/neondatabase/neon/actions/workflows/update_build-tools-image.yml,
```
Comment (bayandin) [inline CONTRIBUTING.md:77]:
```suggestion
You can call it from GitHub UI: https://github.com/neondatabase/neon/actions/workflows/update_build_tools_image.yml,
```

File: .github/workflows/build_and_test.yml [hunk 14]
```diff
@@ -63,22 +63,59 @@ jobs:
...
+
+  build-buildtools-image:
+    needs: [ check-permissions , check-if-build-tools-dockerfile-changed, tag]
+    uses: ./.github/workflows/build_and_push_docker_image.yml
+    if: needs.check-if-build-tools-dockerfile-changed.outputs.docker_file_changed == 'true'
+    with:
+      dockerfile-path: Dockerfile.buildtools
+      image-name: build-tools
+      tag: ${{ needs.tag.outputs.build-tools-tag}}
+    secrets: inherit
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:112]: If this job is skipped (when there are no changes in the dockerfile), won't this cause it to be skipped for all the dependent jobs?
Let's test it as well

File: .github/workflows/update_build-tools-image.yml (hunk 12 above)
Comment (abhigets) [reply .github/workflows/update_build-tools-image.yml]: already pushed the changes

File: .github/workflows/build_and_test.yml (hunk 14 above)
Comment (abhigets) [reply .github/workflows/build_and_test.yml:115]: yep, you are right. thats a good catch. i will look into this

File: .github/workflows/update_build-tools-image.yml
Comment (abhigets) [reply .github/workflows/update_build-tools-image.yml:1]: done

File: .github/workflows/build_and_test.yml (hunk 14 above)
Comment (abhigets) [reply .github/workflows/build_and_test.yml:113]: fixed

File: .github/workflows/build_and_test.yml [hunk 15]
```diff
@@ -175,10 +215,11 @@ jobs:
         run: cargo deny check --hide-inclusion-graph
 
   build-neon:
-    needs: [ check-permissions, tag ]
+    if: ${{ always() }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:218]: Just `always()` maybe a bit too much, we don't want to run these jobs if something has failed or has been cancelled, I think an alternation like this should do the trick

```suggestion
    if: always() && contains(fromJSON('["success", "skipped"]'), needs.build-buildtools-image.result)
```

A side note, I don't really like how it looks, complicates the workflow, and how fragile it is, but this is the only way I know. I hope we can improve in the future

File: .github/workflows/build_and_test.yml [hunk 16]
```diff
@@ -74,11 +74,50 @@ jobs:
...
+        run: |
+          updated_files=$(gh pr --repo neondatabase/neon diff ${{ github.event.pull_request.number }} --name-only)
+          if [[ $updated_files == *"Dockerfile.buildtools"* ]]; then
+            echo "docker_file_changed=true" >> $GITHUB_OUTPUT
+          fi
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  build-buildtools-image:
+    needs: [ check-permissions , check-if-build-tools-dockerfile-changed, tag]
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:106]:
```suggestion
    needs: [ check-permissions, check-if-build-tools-dockerfile-changed, tag]
```

**Review by bayandin — COMMENTED** [review]
Almost there, just a couple of comments left! 🎉

File: .github/workflows/build_and_test.yml [hunk 17]
```diff
@@ -408,10 +449,11 @@ jobs:
         uses: ./.github/actions/save-coverage-data
 
   regress-tests:
+    if: ${{ always() }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:452]:
```suggestion
    if: always() && contains(fromJSON('["success", "skipped"]'), needs.build-buildtools-image.result)
```

File: .github/workflows/build_and_test.yml [hunk 18]
```diff
@@ -108,10 +147,11 @@ jobs:
         run: poetry run mypy .
 
   check-codestyle-rust:
-    needs: [ check-permissions ]
+    if: always() && contains(fromJSON('["success", "skipped"]'), needs.build-buildtools-image.result)
+    needs: [ check-permissions, build-buildtools-image, tag ]
     runs-on: [ self-hosted, gen3, large ]
     container:
-      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/rust:pinned
+      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{  needs.tag.outputs.build-tools-tag }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:154]:
```suggestion
      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{ needs.tag.outputs.build-tools-tag }}
```

File: .github/workflows/build_and_test.yml [hunk 19]
```diff
@@ -408,10 +449,11 @@ jobs:
         uses: ./.github/actions/save-coverage-data
 
   regress-tests:
+    if: ${{ always() }}
     needs: [ check-permissions, build-neon, tag ]
     runs-on: [ self-hosted, gen3, large ]
     container:
-      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/rust:pinned
+      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{  needs.tag.outputs.build-tools-tag }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:456]:
```suggestion
      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{ needs.tag.outputs.build-tools-tag }}
```

File: .github/workflows/build_and_test.yml [hunk 20]
```diff
@@ -447,10 +489,10 @@ jobs:
         uses: ./.github/actions/save-coverage-data
 
   benchmarks:
-    needs: [ check-permissions, build-neon ]
+    needs: [ check-permissions, build-neon, tag]
     runs-on: [ self-hosted, gen3, small ]
     container:
-      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/rust:pinned
+      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{  needs.tag.outputs.build-tools-tag }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:495]:
```suggestion
      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{ needs.tag.outputs.build-tools-tag }}
```

File: .github/workflows/build_and_test.yml [hunk 21]
```diff
@@ -479,12 +521,12 @@ jobs:
...
 
   create-test-report:
-    needs: [ check-permissions, regress-tests, coverage-report, benchmarks ]
+    needs: [ check-permissions, regress-tests, coverage-report, benchmarks, tag ]
     if: ${{ !cancelled() && contains(fromJSON('["skipped", "success"]'), needs.check-permissions.result) }}
 
     runs-on: [ self-hosted, gen3, small ]
     container:
-      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/rust:pinned
+      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{  needs.tag.outputs.build-tools-tag }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:529]:
```suggestion
      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{ needs.tag.outputs.build-tools-tag }}
```

File: .github/workflows/build_and_test.yml [hunk 22]
```diff
@@ -526,11 +568,10 @@ jobs:
             })
 
   coverage-report:
-    needs: [ check-permissions, regress-tests ]
-
+    needs: [ check-permissions, regress-tests, tag ]
     runs-on: [ self-hosted, gen3, small ]
     container:
-      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/rust:pinned
+      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{  needs.tag.outputs.build-tools-tag }}
```
Comment (bayandin) [inline .github/workflows/build_and_test.yml:574]:
```suggestion
      image: 369495373322.dkr.ecr.eu-central-1.amazonaws.com/build-tools:${{ needs.tag.outputs.build-tools-tag }}
```

**Review by bayandin — COMMENTED** [review]
One more forgotten `if: ${{ always() }}` to fix

**Review by bayandin — COMMENTED** [review]
Good work!

I'll approve the PR (I get automerge to merge it) once I’m near the laptop (just not to break something in `main` and leave)

**Review by bayandin — APPROVED** [review]

**Pull request merged**